packages/query-core/src/timeoutManager.ts

```diff
@@ -0,0 +1,174 @@
+/**
+ * Timeout manager does not support passing arguments to the callback.
+ * (`void` is the argument type inferred by TypeScript's default typings for `setTimeout(cb, number)`)
+ */
+export type TimeoutCallback = (_: void) => void
+
+/**
+ * Wrapping `setTimeout` is awkward from a typing perspective because platform
+ * typings may extend the return type of `setTimeout`. For example, NodeJS
+ * typings add `NodeJS.Timeout`; but a non-default `timeoutManager` may not be
+ * able to return such a type.
+ *
+ * Still, we can downlevel `NodeJS.Timeout` to `number` as it implements
+ * Symbol.toPrimitive.
+ */
+export type TimeoutProviderId = number | { [Symbol.toPrimitive]: () => number }
+
+/**
+ * Backend for timer functions.
+ */
+export type TimeoutProvider = {
+  /** Used in error messages. */
+  readonly name: string
+
+  readonly setTimeout: (
+    callback: TimeoutCallback,
+    delay: number,
+  ) => TimeoutProviderId
+  readonly clearTimeout: (timeoutId: number | undefined) => void
+
+  readonly setInterval: (
+    callback: TimeoutCallback,
+    delay: number,
+  ) => TimeoutProviderId
+  readonly clearInterval: (intervalId: number | undefined) => void
+}
+
+export const defaultTimeoutProvider: TimeoutProvider = {
+  name: 'default',
```

Comment:
not sure if we need the name just for the error

Reply:
Does this comment mean "please remove this"?